src/main/java/org/primeframework/mvc/security/cookiesession/UserIDCookieSessionSecurityContext.java

```diff
@@ -0,0 +1,268 @@
+    this.request = request;
+    this.response = response;
+    this.encryptor = encryptor;
+    this.objectMapper = objectMapper;
+    this.clock = clock;
+    this.sessionMaxAge = sessionMaxAge;
+    this.sessionTimeout = sessionTimeout;
+    this.sessionContainerFactory = sessionContainerFactory;
+    this.sessionClass = sessionClass;
+    long timeoutInSeconds = sessionTimeout.toSeconds();
+    this.sessionCookie = new CookieProxy(getCookieName(), timeoutInSeconds, Cookie.SameSite.Strict);
+  }
+
+  protected String getCookieName() {
+    return UserKey;
+  }
+
+  enum CookieExtendResult {
+    // no extension needed
+    Keep,
+    // extend the cookie with the same value, but longer expiration time
+    Extend,
+    // the cookie is now invalid
+    Invalid
+  }
+
+  /**
+   * Figure out whether to extend cookie
+   *
+   * @param signInTime when user signed in originally
+   * @return correct cookie behavior
+   */
+  CookieExtendResult shouldExtendCookie(ZonedDateTime signInTime) {
+    var now = ZonedDateTime.now(clock);
+    var maxSessionAgeTime = signInTime.plus(sessionMaxAge);
+    if (now.isAfter(maxSessionAgeTime)) {
+      return CookieExtendResult.Invalid;
+    }
+    var extensionTime = now.plus(sessionTimeout);
+    if (extensionTime.isAfter(maxSessionAgeTime)) {
+      return CookieExtendResult.Keep;
+    }
+    var halfWayPoint = signInTime.plusMinutes(sessionTimeout.toMinutes() / 2);
+    return now.isAfter(halfWayPoint) ? CookieExtendResult.Extend : CookieExtendResult.Keep;
+  }
+
+  /**
+   * Get the currently logged in user. Only calls retrieveUserById once per request cycle.
+   *
+   * @return null if no user is logged in. Otherwise the user object as retrieved by retrieveUserById using the ID from the session
+   */
+  @Override
+  public Object getCurrentUser() {
+    SerializedSessionContainer sessionContainer = getSessionContainer();
+    if (sessionContainer == null) {
+      return null;
+    }
+    if (sessionContainer instanceof HydratedSessionContainer inMemory) {
+      return inMemory.user();
+    }
+    var user = retrieveUserById(sessionContainer.userId());
+    var newContainer = sessionContainerFactory.createHydrated(sessionContainer, user);
+    this.request.setAttribute(UserKey, newContainer);
+    return newContainer.user();
+  }
+
+  /**
+   * Get existing deserialized session container from cookie
+   *
+   * @return value from cookie, null if no existing session
+   */
+  private SerializedSessionContainer getSessionContainer() {
+    var container = (SerializedSessionContainer) this.request.getAttribute(UserKey);
+    if (container != null) {
+      return container;
+    }
+    var cookie = this.sessionCookie.get(this.request);
+    if (cookie == null) {
+      return null;
+    }
+    try {
+      container = CookieTools.fromJSONCookie(cookie, sessionClass, true, encryptor, objectMapper);
```

Comment:
@robotdan please see `SessionContainerFactory` too, but is there a more elegant way with Guice to handle this? You can see how the `MockUserIDCookieSessionSecurityContext` test class deals with this.

The only other thing I could think of to optimize this without getting hairy is to go down the route of putting

`@JsonTypeInfo(include = As.WRAPPER_OBJECT, use = Id.MINIMAL_CLASS)`

on the interface. But that then puts the (still encrypted) browser payload in charge of choosing its deserialization class, which is probably slightly subpar to the way I have it now. Thoughts?

Reply:
Commits since last review:

1a634ba5aa24b51934d3fc8091370280570783d8

f95f2b44575dbc47f5a110ce9209af192d0b2700

ffa174350b81ba8584976a6c80e1c4b569340b21